android: Input mapping

t895 opened the pull request:
Input mapping is finally here and we have a lot to cover!

Here's what's new
- Connecting and disconnecting each controller
- Changing each controller's type between Pro Controller, Left Joycon, Right Joycon, Handheld (player 1 only), and Gamecube
- Selecting a device from the "Input device" dropdown to auto-map a controller
- If you don't have any controllers mapped, you can still start any game and have the first connected controller ready to play with auto-mapping
- Creating/loading input profiles
- Mapping each control with an axis or button
- Each mapping has options for modifiers like invert, toggle, and turbo
- Each stick has the option to set a deadzone and a range
- Controllers and phones (system vibration motors generally) can vibrate now

Limitations and potential problems
- HD Rumble is not supported yet
- Only gamepads can be mapped so don't expect to hook up a keyboard or play Mario with your volume buttons
- There is no controller visualization page yet
- Per-game input profiles were not included due to a longstanding bug with profile config. This will be addressed in a later PR.
- If there are any inconsistency issues with auto mapping, please report them to me. The expected layout for face buttons should be A - Right button, B - Bottom button, X - Top button, Y - Left button.
- A note for **AYN Odin 2** users: There is a bug affecting the Odin's controller handling where the first controller connected besides the integrated one will share the same ID information as the integrated controller and have broken vibrations/auto-mapping (but controllers connected past that will be fine). This is known to AYN and cannot be fixed on my end so please do not report this.

I'd also like to thank @german77 for laying some of the groundwork in his input engine that I needed to get started.

Closes
#12841
#12197
#12121
#12057
#11593
#11124

<details>
<summary>Media</summary>

I've decided to spare your eyes this time with dark mode

![image](https://github.com/yuzu-emu/yuzu/assets/14132249/16bbea98-bab4-42ca-b055-f09e063cef1f)

![image](https://github.com/yuzu-emu/yuzu/assets/14132249/6988022d-15bb-47a4-bfb3-78b2a4876f8c)

![image](https://github.com/yuzu-emu/yuzu/assets/14132249/6ea43979-6e85-4097-88f1-bd23e03228cc)

I'm really hoping that everyone gets the message about the input device drop down -_-

![anim-single-dir](https://github.com/yuzu-emu/yuzu/assets/14132249/63b97a91-09db-4f49-a96e-9a5a9442aaa9)

![anim-two-dir](https://github.com/yuzu-emu/yuzu/assets/14132249/7578bffd-78cf-468b-acb7-dc42200bbb86)

Some pretty animations to enjoy while you're mapping :)

![image](https://github.com/yuzu-emu/yuzu/assets/14132249/388949be-cafa-45d3-abc1-843d1655cba5)

Go crazy with the modifiers

![image](https://github.com/yuzu-emu/yuzu/assets/14132249/a6485098-4f4a-442b-94ef-e66486f518d1)

It's about time that we got a deadzone slider!

![image](https://github.com/yuzu-emu/yuzu/assets/14132249/86c6d0c6-7e36-4958-990e-706bfb6051f3)

![image](https://github.com/yuzu-emu/yuzu/assets/14132249/6790ec36-f603-4900-9a5d-b7cb2822c9c1)

Love you Golden ❤️
</details>

german77 commented on src/input_common/drivers/android.h:
```diff
@@ -43,12 +46,86 @@ class Android final : public InputEngine {
-    PadIdentifier GetIdentifier(std::size_t controller_number) const;
+    PadIdentifier GetIdentifier(const std::string& guid, size_t port) const;
+
+    const s32 AXIS_X = 0;
```
static constexpr

german77 commented on src/input_common/drivers/android.h:
```diff
@@ -43,12 +46,86 @@ class Android final : public InputEngine {
                         float gyro_y, float gyro_z, float accel_x, float accel_y, float accel_z);
 
+    Common::Input::DriverResult SetVibration(
+        [[maybe_unused]] const PadIdentifier& identifier,
```
headers don't need the [[maybe_unused]] attribute

german77 commented on src/input_common/drivers/android.cpp:
```diff
@@ -37,10 +54,295 @@ void Android::SetMotionState(std::size_t controller_number, u64 delta_timestamp,
+    return Common::Input::DriverResult::NotSupported;
+}
+
+bool Android::IsVibrationEnabled([[maybe_unused]] const PadIdentifier& identifier) {
```
identifier is actually used

german77 commented on src/input_common/drivers/android.cpp:
```diff
@@ -37,10 +54,295 @@ void Android::SetMotionState(std::size_t controller_number, u64 delta_timestamp,
 
-PadIdentifier Android::GetIdentifier(std::size_t controller_number) const {
+Common::Input::DriverResult Android::SetVibration(
+    [[maybe_unused]] const PadIdentifier& identifier,
```
identifier and vibration are used

german77 commented on src/hid_core/frontend/emulated_controller.cpp:
```diff
@@ -1277,6 +1283,12 @@ bool EmulatedController::SetVibration(DeviceIndex device_index, const VibrationV
+
+    // Send vibrations to Android's input overlay
+    if (npad_id_type == NpadIdType::Handheld || npad_id_type == NpadIdType::Player1) {
+        output_devices[4]->SetVibration(status);
```
Forgot this check handled && player1 isn't needed here since the params are already defined just for handheld and player1. So we can leave this part as just `output_devices[4]->SetVibration(status);`

german77 commented on src/common/settings_input.h:
```diff
@@ -395,6 +395,10 @@ struct PlayerInput {
     std::string profile_name;
+
+#ifdef ANDROID
+    bool use_system_vibrator;
```
Do we need to exclude this from other systems? I would rather just leave a comment saying this is for android support.

german77 commented on src/android/app/src/main/java/org/yuzu/yuzu_emu/features/settings/ui/SettingsActivity.kt:
```diff
@@ -137,6 +137,8 @@ class SettingsActivity : AppCompatActivity() {
         super.onStop()
         Log.info("[SettingsActivity] Settings activity stopping. Saving settings to INI...")
         if (isFinishing) {
+            NativeInput.unloadInputProfiles()
```
unload isn't required here. You can reload directly.

german77 reviewed (commented):
I like how it was integrated with the existing input backend

t895 commented:
Reworked the "Input Device" setting
It's now split into two options to make things clearer.
- Auto-map a controller: Select a device from the list to attempt auto mapping with
- Input mapping filter: Select a device or "Any" from the list to use as a filter for the mapping dialogs

dannygaray60 commented:
@t895 thanks for yoy work! Very appreciated, just as an advice, the Xbox Series X controller still has the layout inverted when you use the automatic mapping

t895 commented:
Are A and B flipped or A and B + X and Y flipped?

dannygaray60 commented:
@t895 the A/B are fliped and X/Y fliped

H0nzik1 commented:
@t895 could you please verify why it is not working in Rayman Legends Definitive Edition? I have mapped 3 controllers, no luck. I am still using this fork https://github.com/chulinx/yuzu-android/releases to play it. Thanks

darix080 commented:
I am not able to remap my controller, could you help me?

This is my controller
![Screenshot_20240207_000724_Gamepad Tester](https://github.com/yuzu-emu/yuzu/assets/159198668/2456051a-f1a2-48a2-a98b-b13a646201ff)

JomPyx commented:
Thanks for adding the remapping button. But I hope it will be fixed for generic gamepad usb. Because the button A, B, X, Y, R, L, ZR, ZL cannot be input and only the analog, dpad is working. I hope you will fix it soon. 🙏🙏
